### PR TITLE
WIP fix(notifications): more reliable notifications for ACCESS_LOGGED_IN entities

### LINKED ISSUE #10133: ACCESS_LOGGED_IN and has_access_to_entity()
We have a flaw in our access check that manifests in notifications handling. If an item is saved with `ACCESS_LOGGED_IN`, user never gets notified about it because the SQL query in `has_access_to_entity()` does not include 1.

### PR DESCRIPTION
**Pull request by mrclay**
`has_access_to_entity()` now sees all public and logged-in-only entities as accessible by all users.

Fixes #10133

### CODE REVIEW

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 1]
```diff
@@ -317,7 +317,13 @@ function getWhereSql(array $options = array()) {
 	 * @return bool
 	 */
 	function hasAccessToEntity($entity, $user = null) {
-		
+		if (!$entity instanceof \ElggEntity) {
+			return false;
+		}
+
+		if (in_array($entity->access_id, [ACCESS_PUBLIC, ACCESS_LOGGED_IN])) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Only if there is a user (logged in or passed with attributes).

File: engine/tests/ElggCoreAccessSQLTest.php [hunk 2]
```diff
@@ -175,6 +175,49 @@ public function testAccessPluginHookAddAnd() {
...
+		$session->removeLoggedInUser();
+		$this->assertFalse(has_access_to_entity($object));
+		$this->assertFalse(has_access_to_entity($object, $this->user));
+		$session->setLoggedInUser($test_user);
+
+		$object->access_id = ACCESS_PUBLIC;
+		$object->save();
+
+		$session->removeLoggedInUser();
+		$this->assertTrue(has_access_to_entity($object));
```
Comment (hypeJunction) [inline engine/tests/ElggCoreAccessSQLTest.php]: This should be false

Comment (mrclay) [inline engine/tests/ElggCoreAccessSQLTest.php]: You mean line 203 should be false, right? For this one the entity is public.

**Comment by mrclay** [comment]
All these tests pass without any changes.

File: engine/tests/ElggCoreAccessSQLTest.php (hunk 2 above)
Comment (hypeJunction) [inline engine/tests/ElggCoreAccessSQLTest.php]: Yes, sorry

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 3]
```diff
@@ -317,12 +317,22 @@ function getWhereSql(array $options = array()) {
...
+
+//		if ($entity->access_id == ACCESS_PUBLIC) {
+//			return true;
+//		}
+
 		// See #7159. Must not allow ignore access to affect query
 		$ia = elgg_set_ignore_access(false);
 	
 		if (!isset($user)) {
+//			if (elgg_is_logged_in() && $entity->access_id == ACCESS_LOGGED_IN) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Shouldn't use session check. Just if there is a user

Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Never mind. Being stupid.

**Comment by mrclay** [comment]
All tests are already passing, so are you sure 1.12 is affected?

**Comment by hypeJunction** [comment]
No, not sure. I thought this function didn't change since a while

**Comment by hypeJunction** [comment]
So, this should go in 2.2?

**Comment by mrclay** [comment]
I guess we could add the test in 1.12 and if it fails in 2.2 we fix there?

**Comment by hypeJunction** [comment]
I just made the 1.12 release. Going to make 2.2 now. Do you want to squeeze this in?

**Comment by mrclay** [comment]
I may not get to this today.

**Comment by hypeJunction** [comment]
Argh. I accidentally committed 2.2 config to 1.12 branch... Will have to redo...

**Comment by hypeJunction** [comment]
Replaced by #10175

**Pull request closed without merging**